File: manager/src/main/java/moe/shizuku/manager/SettingsFragment.java
package moe.shizuku.manager;

import android.content.Context;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Locale;

import moe.shizuku.manager.app.ThemeHelper;
import moe.shizuku.manager.viewmodel.AppsViewModel;
import moe.shizuku.manager.viewmodel.SharedViewModelProviders;
import moe.shizuku.preference.ListPreference;
import moe.shizuku.preference.Preference;
import moe.shizuku.preference.PreferenceFragment;
import moe.shizuku.preference.SwitchPreference;
import rikka.core.util.ResourceUtils;
import rikka.html.text.HtmlCompat;
import rikka.material.app.DayNightDelegate;
import rikka.material.app.LocaleDelegate;
import rikka.recyclerview.RecyclerViewHelper;

public class SettingsFragment extends PreferenceFragment {

    public static final String KEY_LANGUAGE = ShizukuManagerSettings.LANGUAGE;
    public static final String KEY_NIGHT_MODE = ShizukuManagerSettings.NIGHT_MODE;
    public static final String KEY_BLACK_NIGHT_THEME = ThemeHelper.KEY_BLACK_NIGHT_THEME;
    public static final String KEY_NO_V2 = ShizukuManagerSettings.NO_V2;

    private ListPreference languagePreference;
    private Preference nightModePreference;
    private SwitchPreference blackNightThemePreference;
    private SwitchPreference noV2Preference;

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        getPreferenceManager().setStorageDeviceProtected();
        getPreferenceManager().setSharedPreferencesName(ShizukuManagerSettings.NAME);
        getPreferenceManager().setSharedPreferencesMode(Context.MODE_PRIVATE);

        setPreferencesFromResource(R.xml.settings, null);

        languagePreference = (ListPreference) findPreference(KEY_LANGUAGE);
        nightModePreference = findPreference(KEY_NIGHT_MODE);
        blackNightThemePreference = (SwitchPreference) findPreference(KEY_BLACK_NIGHT_THEME);
        noV2Preference = (SwitchPreference) findPreference(KEY_NO_V2);

        AppsViewModel viewModel = SharedViewModelProviders.of(this).get("apps", AppsViewModel.class);
        viewModel.getPackages().observe(this, object -> {
            if (object.data != null)
                updateData(object.data);
        });
        if (viewModel.getPackages() != null
                && viewModel.getPackages().getValue() != null
                && viewModel.getPackages().getValue().data != null) {
            updateData(viewModel.getPackages().getValue().data);
        }

        languagePreference.setOnPreferenceChangeListener((preference, newValue) -> {
            if (newValue instanceof String) {
                Locale locale;
                if ("SYSTEM".equals(newValue)) {
                    locale = LocaleDelegate.getSystemLocale();
                } else {
                    locale = Locale.forLanguageTag((String) newValue);
                }
                LocaleDelegate.setDefaultLocale(locale);
                recreateActivity();
            }
            return true;
        });

        String tag = languagePreference.getValue();
        int index = Arrays.asList(languagePreference.getEntryValues()).indexOf(tag);

        List<String> localeName = new ArrayList<>();
        List<String> localeNameUser = new ArrayList<>();
        Locale userLocale = ShizukuManagerSettings.getLocale();
        for (int i = 1; i < languagePreference.getEntries().length; i++) {
            Locale locale = Locale.forLanguageTag(languagePreference.getEntries()[i].toString());
            localeName.add(!TextUtils.isEmpty(locale.getScript()) ? locale.getDisplayScript(locale) : locale.getDisplayName(locale));
            localeNameUser.add(!TextUtils.isEmpty(locale.getScript()) ? locale.getDisplayScript(userLocale) : locale.getDisplayName(userLocale));
        }
        for (int i = 1; i < languagePreference.getEntries().length; i++) {
            if (index != i) {
                languagePreference.getEntries()[i] = HtmlCompat.fromHtml(
                        String.format("%s - %s",
                                localeName.get(i - 1),
                                localeNameUser.get(i - 1)
                        ));
            } else {
                languagePreference.getEntries()[i] = localeNameUser.get(i - 1);
            }
        }

        if (TextUtils.isEmpty(tag) || "SYSTEM".equals(tag)) {
            languagePreference.setSummary(getString(R.string.follow_system));
        } else if (index != -1) {
            String name = localeNameUser.get(index - 1);
            languagePreference.setSummary(name);
        }

        nightModePreference.setOnPreferenceChangeListener((preference, o) -> {
            if (o instanceof Integer) {
                int mode = (int) o;
                if (ShizukuManagerSettings.getNightMode() != mode) {
                    DayNightDelegate.setDefaultNightMode(mode);
                    recreateActivity();
                }
            }
            return true;
        });

        blackNightThemePreference.setOnPreferenceChangeListener(((preference, newValue) -> {
            if (ResourceUtils.isNightMode(requireContext().getResources().getConfiguration()))
                recreateActivity();
            return true;
        }));
    }

    @Nullable
    @Override
    public DividerDecoration onCreateItemDecoration() {
        return new CategoryDivideDividerDecoration();
    }

    @Override
    public RecyclerView onCreateRecyclerView(LayoutInflater inflater, ViewGroup parent, Bundle savedInstanceState) {
        RecyclerView recyclerView = super.onCreateRecyclerView(inflater, parent, savedInstanceState);
        RecyclerViewHelper.fixOverScroll(recyclerView);

        int padding = (int) (8 * recyclerView.getContext().getResources().getDisplayMetrics().density);
        recyclerView.setPaddingRelative(recyclerView.getPaddingStart(), 0, recyclerView.getPaddingEnd(), padding);
        ViewGroup.LayoutParams _lp = recyclerView.getLayoutParams();
        if (_lp instanceof FrameLayout.LayoutParams) {
            FrameLayout.LayoutParams lp = (FrameLayout.LayoutParams) _lp;
            lp.leftMargin = lp.rightMargin = (int) recyclerView.getContext().getResources().getDimension(R.dimen.activity_horizontal_margin);
        }
        return recyclerView;
    }

    private void recreateActivity() {
        if (getActivity() != null) {
            getActivity().recreate();
        }
    }

    private void updateData(List<PackageInfo> data) {
        int count = 0;
        for (PackageInfo pi : data) {
            ApplicationInfo ai = pi.applicationInfo;
            if (ai.metaData == null || !ai.metaData.getBoolean("moe.shizuku.client.V3_SUPPORT")) {
                count++;
            }
        }

        noV2Preference.setSummary(requireContext().getResources().getQuantityString(R.plurals.start_legacy_service_summary, count, count));
    }
}

File: manager/src/main/java/moe/shizuku/manager/SettingsFragment.kt
package moe.shizuku.manager

import android.content.Context
import android.content.pm.PackageInfo
import android.os.Bundle
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.FrameLayout
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.RecyclerView
import moe.shizuku.manager.app.ThemeHelper
import moe.shizuku.manager.viewmodel.AppsViewModel
import moe.shizuku.manager.viewmodel.DataWrapper
import moe.shizuku.manager.viewmodel.SharedViewModelProviders
import moe.shizuku.preference.*
import rikka.core.util.ResourceUtils
import rikka.html.text.HtmlCompat
import rikka.material.app.DayNightDelegate
import rikka.material.app.LocaleDelegate
import rikka.material.widget.*
import rikka.recyclerview.RecyclerViewHelper
import java.util.*

class SettingsFragment : PreferenceFragment() {

    companion object {

        const val KEY_LANGUAGE = ShizukuManagerSettings.LANGUAGE
        const val KEY_NIGHT_MODE = ShizukuManagerSettings.NIGHT_MODE
        const val KEY_BLACK_NIGHT_THEME = ThemeHelper.KEY_BLACK_NIGHT_THEME
        const val KEY_NO_V2 = ShizukuManagerSettings.NO_V2

        init {
            SimpleMenuPreference.setLightFixEnabled(true)
        }
    }

    private var languagePreference: ListPreference? = null
    private var nightModePreference: Preference? = null
    private var blackNightThemePreference: SwitchPreference? = null
    private var noV2Preference: SwitchPreference? = null

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        preferenceManager.setStorageDeviceProtected()
        preferenceManager.sharedPreferencesName = ShizukuManagerSettings.NAME
        preferenceManager.sharedPreferencesMode = Context.MODE_PRIVATE
        setPreferencesFromResource(R.xml.settings, null)

        languagePreference = findPreference(KEY_LANGUAGE) as ListPreference
        nightModePreference = findPreference(KEY_NIGHT_MODE)
        blackNightThemePreference = findPreference(KEY_BLACK_NIGHT_THEME) as SwitchPreference
        noV2Preference = findPreference(KEY_NO_V2) as SwitchPreference

        val viewModel = SharedViewModelProviders.of(this).get("apps", AppsViewModel::class.java)
        viewModel.packages.observe(this, Observer { `object`: DataWrapper<List<PackageInfo>?> -> if (`object`.data != null) updateData(`object`.data) })
        if (viewModel.packages != null && viewModel.packages.value != null && viewModel.packages.value!!.data != null) {
            updateData(viewModel.packages.value!!.data)
        }

        languagePreference!!.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { _: Preference?, newValue: Any ->
            if (newValue is String) {
                val locale: Locale = if ("SYSTEM" == newValue) {
                    LocaleDelegate.systemLocale
                } else {
                    Locale.forLanguageTag(newValue)
                }
                LocaleDelegate.defaultLocale = locale
                activity?.recreate()
            }
            true
        }

        val tag = languagePreference!!.value
        val index = listOf(*languagePreference!!.entryValues).indexOf(tag)
        val localeName: MutableList<String> = ArrayList()
        val localeNameUser: MutableList<String> = ArrayList()
        val userLocale = ShizukuManagerSettings.getLocale()
        for (i in 1 until languagePreference!!.entries.size) {
            val locale = Locale.forLanguageTag(languagePreference!!.entries[i].toString())
            localeName.add(if (!TextUtils.isEmpty(locale.script)) locale.getDisplayScript(locale) else locale.getDisplayName(locale))
            localeNameUser.add(if (!TextUtils.isEmpty(locale.script)) locale.getDisplayScript(userLocale) else locale.getDisplayName(userLocale))
        }

        for (i in 1 until languagePreference!!.entries.size) {
            if (index != i) {
                languagePreference!!.entries[i] = HtmlCompat.fromHtml(String.format("%s - %s",
                        localeName[i - 1],
                        localeNameUser[i - 1]
                ))
            } else {
                languagePreference!!.entries[i] = localeNameUser[i - 1]
            }
        }

        if (TextUtils.isEmpty(tag) || "SYSTEM" == tag) {
            languagePreference!!.summary = getString(R.string.follow_system)
        } else if (index != -1) {
            val name = localeNameUser[index - 1]
            languagePreference!!.summary = name
        }
        nightModePreference!!.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { _: Preference?, value: Any? ->
            if (value is Int) {
                if (ShizukuManagerSettings.getNightMode() != value) {
                    DayNightDelegate.setDefaultNightMode(value)
                    activity?.recreate()
                }
            }
            true
        }
        blackNightThemePreference!!.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { _: Preference?, _: Any? ->
            if (ResourceUtils.isNightMode(requireContext().resources.configuration)) activity?.recreate()
            true
        }
    }

    override fun onCreateItemDecoration(): DividerDecoration? {
        return CategoryDivideDividerDecoration()
    }

    override fun onCreateRecyclerView(inflater: LayoutInflater, parent: ViewGroup, savedInstanceState: Bundle?): RecyclerView {
        val recyclerView = super.onCreateRecyclerView(inflater, parent, savedInstanceState) as BorderRecyclerView
        RecyclerViewHelper.fixOverScroll(recyclerView)
        val padding = (8 * recyclerView.context.resources.displayMetrics.density).toInt()
        recyclerView.setPaddingRelative(recyclerView.paddingStart, 0, recyclerView.paddingEnd, padding)
        val lp = recyclerView.layoutParams
        if (lp is FrameLayout.LayoutParams) {
            lp.rightMargin = recyclerView.context.resources.getDimension(R.dimen.rd_activity_horizontal_margin).toInt()
            lp.leftMargin = lp.rightMargin
        }
        val verticalPadding = resources.getDimension(R.dimen.list_vertical_padding).toInt()
        recyclerView.setInitialPadding(
                recyclerView.initialPaddingLeft,
                recyclerView.initialPaddingTop,
                recyclerView.initialPaddingRight,
                recyclerView.initialPaddingBottom + verticalPadding
        )
        recyclerView.borderViewDelegate.borderVisibilityChangedListener = BorderView.OnBorderVisibilityChangedListener { top: Boolean, _: Boolean, _: Boolean, _: Boolean -> (activity as SettingsActivity?)?.appBar?.setRaised(!top) }
        return recyclerView
    }

    private fun updateData(data: List<PackageInfo>?) {
        var count = 0
        for (pi in data!!) {
            val ai = pi.applicationInfo
            if (ai.metaData == null || !ai.metaData.getBoolean("moe.shizuku.client.V3_SUPPORT")) {
                count++
            }
        }
        noV2Preference!!.summary = requireContext().resources.getQuantityString(R.plurals.start_legacy_service_summary, count, count)
    }
}